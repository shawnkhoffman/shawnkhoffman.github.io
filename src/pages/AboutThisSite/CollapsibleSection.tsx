import React, { useState, useId } from 'react';

interface CollapsibleSectionProps {
  title: string;
  content: string;
  testId?: string;
}

const CollapsibleSection: React.FC<CollapsibleSectionProps> = ({ 
  title, 
  content,
  testId = 'collapsible-section'
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const uniqueId = useId();
  const contentId = `collapsible-content-${uniqueId}`;

  const handleToggle = (event: React.SyntheticEvent) => {
    event.preventDefault();
    setIsOpen(prev => !prev);
  };

  const handleKeyDown = (event: React.KeyboardEvent) => {
    if (event.key === 'Enter' || event.key === ' ') {
      event.preventDefault();
      handleToggle(event);
    }
  };

  const handleTouchEnd = (event: React.TouchEvent) => {
    event.preventDefault();
    handleToggle(event);
  };

  return (
    <div
      data-testid={testId}
      className={`collapse collapse-arrow bg-base-200 ${isOpen ? 'collapse-open' : ''}`}
    >
      <button
        type="button"
        className="collapse-title text-xl font-medium"
        aria-expanded={isOpen}
        aria-controls={contentId}
        onClick={handleToggle}
        onKeyDown={handleKeyDown}
        onTouchEnd={handleTouchEnd}
        tabIndex={0}
      >
        {title}
      </button>
      <div
        id={contentId}
        className="collapse-content"
        role="region"
        aria-hidden={!isOpen}
      >
        <p>{content}</p>
      </div>
    </div>
  );
};

export default CollapsibleSection;
