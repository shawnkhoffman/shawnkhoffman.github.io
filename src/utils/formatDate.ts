export const formatDate = (date: string | Date): string => {
    const options: Intl.DateTimeFormatOptions = { year: 'numeric', month: 'long', day: 'numeric' };
    const formattedDate = new Date(date).toLocaleDateString(undefined, options);
    return formattedDate;
  };