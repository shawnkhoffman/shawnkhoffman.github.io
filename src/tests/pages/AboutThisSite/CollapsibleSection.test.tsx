import { fireEvent, cleanup, within } from '@testing-library/react';
import { describe, it, expect, afterEach } from 'vitest';
import CollapsibleSection from '../../../pages/AboutThisSite/CollapsibleSection';
import { render } from '../../utils/test-utils';
import { TEST_IDS } from '../../utils/test-constants';

describe('CollapsibleSection', () => {
  const mockProps = {
    title: 'Section Title',
    content: 'This is the collapsible content.',
    testId: TEST_IDS.COLLAPSIBLE_SECTION
  };

  afterEach(() => {
    cleanup();
  });

  describe('Rendering', () => {
    it('matches snapshot', () => {
      const { container } = render(<CollapsibleSection {...mockProps} />, { withLayout: false });
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Structure', () => {
    it('contains required elements', () => {
      const { container } = render(<CollapsibleSection {...mockProps} />, { withLayout: false });
      const section = container.firstChild as HTMLElement;
      const titleElement = within(section).getByText(mockProps.title);
      const contentElement = within(section).getByText(mockProps.content);
      
      expect(titleElement).toBeInTheDocument();
      expect(contentElement).toBeInTheDocument();
    });
  });

  describe('Interactions', () => {
    it('toggles content visibility on click', () => {
      const { container } = render(<CollapsibleSection {...mockProps} />, { withLayout: false });
      const section = container.firstChild as HTMLElement;
      const button = within(section).getByRole('button');
      
      fireEvent.click(button);
      expect(button).toHaveAttribute('aria-expanded', 'true');
      
      fireEvent.click(button);
      expect(button).toHaveAttribute('aria-expanded', 'false');
    });
  });

  describe('Accessibility', () => {
    it('maintains correct ARIA states', () => {
      const { container } = render(<CollapsibleSection {...mockProps} />, { withLayout: false });
      const section = container.firstChild as HTMLElement;
      const button = within(section).getByRole('button');
      const content = section.querySelector('[role="region"]');
      
      expect(button).toHaveAttribute('aria-expanded', 'false');
      expect(content).toHaveAttribute('aria-hidden', 'true');
      
      fireEvent.click(button);
      
      expect(button).toHaveAttribute('aria-expanded', 'true');
      expect(content).toHaveAttribute('aria-hidden', 'false');
    });

    it('maintains tab order', () => {
      const { container } = render(<CollapsibleSection {...mockProps} />, { withLayout: false });
      const section = container.firstChild as HTMLElement;
      const button = within(section).getByRole('button');
      
      expect(button).toHaveAttribute('tabindex', '0');
    });

    it('supports screen readers', () => {
      const { container } = render(<CollapsibleSection {...mockProps} />, { withLayout: false });
      const section = container.firstChild as HTMLElement;
      const button = within(section).getByRole('button');
      const content = section.querySelector('[role="region"]');
      
      expect(button).toHaveAttribute('aria-controls');
      expect(content).toHaveAttribute('role', 'region');
    });
  });
});
