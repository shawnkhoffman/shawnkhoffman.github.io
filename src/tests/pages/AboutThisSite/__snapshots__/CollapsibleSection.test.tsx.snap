// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CollapsibleSection > Accessibility > maintains tab order 1`] = `
<div
  class="collapse collapse-arrow bg-base-200 "
  data-testid="collapsible-section"
>
  <button
    aria-controls="collapsible-content-:r0:"
    aria-expanded="false"
    class="collapse-title text-xl font-medium"
    tabindex="0"
    type="button"
  >
    Section Title
  </button>
  <div
    aria-hidden="true"
    class="collapse-content"
    id="collapsible-content-:r0:"
    role="region"
  >
    <p>
      This is the collapsible content.
    </p>
  </div>
</div>
`;
